### ISSUE TITLE
Memory map block with other overlay blocks

### ISSUE THREAD

**Issue by gamelaster** [NationalSecurityAgency/ghidra#4772]
I have block `TCM` (Tight Cache Memory), firmware uses this area for stack, gp, few functions, data etc. I mapped the whole TCM area (as uninitialized), so stack/gp references works properly.

As I mentioned, it contains few functions, which are copied from Flash to TCM. Normally I would create another memory block, and map Flash area to the TCM area, but it collides with TCM block (which makes sense). So I can use "Overlay" feature, but in overlay I can't disassemble the code, and decompiler doesn't care about overlays at all, so it maps to uninitialized area in TCM block. I can remove the TCM area and map only those, but then stack variables don't belong to any memory map.

What is the best way of doing this? I got idea where I can make `bin` file with whole TCM area, and copy parts from Flash into it, and then import it as full memory block, but that's quite hacky. Is there any other way?

Thanks.

**Comment by ghidra1** [comment]
What processor spec (i.e., language ID) are you using?

What version of Ghidra were you using?  There were some overlay regression error introduce with 10.2 that have been corrected in 10.2.2.

Are you conflating overlay memory blocks with byte-mapped memory blocks?  Seems a byte-mapped block would be more applicable to your situation (TCM Functions -> Flash firmware).

The stack space associated with SP (see cspec `stackpointer` element) should always be associated with a physical address space and is never dereferenced to an absolute offset in memory.  Overlays, in general, are used to allow the same memory region by address to have alternative views.  Unfortunately, knowing which overlay to dereference into is limited to decompiling a single function within an overlay and using absolute/relative memory references withing the same overlay.   A function within one overlay cannot reference data/code in another overlay since we have no way to mechanism to identify the other overlay.

In general, indirect dereferencing of register pointers (e.g., GP) do not currently handle dereferencing into an overlay space but should handle byte-mapped memory blocks.  @caheckman may be able to offer more insight into the decompiler's ability to indirectly dereference a constant pointer into the same overlay that contains the function being decompiled - last I was aware of this was not supported.

**Comment by gamelaster** [comment]
Hello @ghidra1 ,
I'm sorry, I forgot to provide more information about project and ghidra version.
Language ID is RISCV:LE:32:RV32GC , Ghidra 10.2.2 (2022-Nov-15 2224 CET).

Thanks for explanation and it makes sense now.
Basically, only thing I wanted to do, is to map whole TCM area as not-initialized block, but inside this TCM block, put data inside, which is copied from Flash by firmware. I thought that overlay is for this (so I have block with not-initialized data, and then by overlay add content to various parts of it).

**Comment by ghidra1** [comment]
I am still puzzled why use an overlay instead of the actual memory region?  Think of overlays as a time-based distinction (i.e., the same address(es) can be dynamically loaded with different content "code" at different times).   If for the purpose of analysis, if the code is loaded from Flash once only to a region why not just copy the bytes of create a byte-mapped block which would map the actual Flash bytes into the memory region where it gets copied.

In some situations, this may involve an unpacker (i.e., decompress) in which case the unpacking must be done manually or via a script and the unpacked bytes copied into a new initialized memory block(s) where they get loaded. 

I would not use an overlay-block unless you have no other choice since it has quite a few limitations since the same memory address has more than one possible block.  This is not the case for normal memory blocks or byte-mapped blocks.

**Comment by ghidra1** [comment]
I am going to close this ticket since I believe everything is working as intended.  Fee free to post here if you need additional clarrification.

**Comment by gamelaster** [comment]
@ghidra1 Yes, most of my questions were answered and so far, I didn't had any issues with it :blush: Thank you very much.
